### PR TITLE
Implement conditional workflow steps

### PR DESCRIPTION
**Pull request by mvdbeek**
This PR implements executing tool and subworkflow steps conditionally, based on the result of a javascript expression. The workflow editor will for now only allow the simplest expression, which is a connection to a boolean parameter (from an expression tool or a workflow input).
An example gxformat2 workflow looks like this:
```yaml
class: GalaxyWorkflow
inputs:
  should_run:
    type: boolean
  some_file:
    type: data
steps:
  cat1:
    tool_id: cat1
    in:
      input1: some_file
      should_run: should_run
    when: $(inputs.should_run)
```
If `when` evaluates to `true` the step will run, if it evaluates to `false` it will not run.
This is also implemented for mapped over tool and workflow steps; if an output is skipped it will be produce a null value. You can think of this as an optional output. That means you can use the `pickValue` expression tool to pick for instance the first created output for two mutually exclusive processing steps.

Includes a whole new infrastructure for displaying scheduling warnings, failures and cancellations:

Screenshot 2023-01-16 at 22 23 19

Remaining work: 
 - Update the workflow editor
 - Some visual indicator for partial collections in the history ?

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by mvdbeek** [comment]
I think the backend is ready for a review ... I would probably split out the UI work into another PR (I should wrap up the first part of the workflow editor overhaul, that would make this easier).

File: lib/galaxy/workflow/modules.py [hunk 1]
```diff
@@ -103,6 +105,99 @@ def __str__(self):
...
+            step_input = step.inputs_by_name.get(key)
+            if step_input and step_input.value_from is not None:
+                value_from_expressions[key] = step_input.value_from
+
+    if not value_from_expressions and when_expression is None:
+        return {}
+
+    hda_references = []
+
+    def to_cwl(value):
```
Comment (jmchilton) [inline lib/galaxy/workflow/modules.py:140]: This is so complicated and we're only basically testing boolean parameters right?

What if an integer is sent - what happens? A dataset? Etc...

Can we have a test that the workflow dies if an invalid expression is set in there?

**Review by jmchilton — APPROVED** [review]

File: lib/galaxy/workflow/modules.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/workflow/modules.py:140]: I've included a bunch of tests and I've also added "invocation messages", see the screenshot in the description ...  I think we're now displaying (almost?) all of the possible workflow errors in the API and UI.

**Review by dannon — APPROVED** [review]

File: lib/galaxy/workflow/modules.py [hunk 2]
```diff
@@ -104,6 +118,126 @@ def __str__(self):
...
+            return value
+
+    def from_cwl(value):
+        # TODO: turn actual files into HDAs here ... somehow I suppose. Things with
+        # file:// locations for instance.
+        if isinstance(value, dict) and "class" in value and "location" in value:
+            if value["class"] == "File":
+                # This is going to re-file -> HDA this each iteration I think, not a good
+                # implementation.
+                return progress.raw_to_galaxy(value)
```
Comment (jmchilton) [inline lib/galaxy/workflow/modules.py:191]: This is nothing right? How did we let this in 😆. It seems it hasn't caused any harm yet!

**Pull request merged**